clang/lib/Serialization/ASTWriter.cpp

```diff
@@ -1824,6 +1824,27 @@ struct InputFileEntry {
   uint32_t ContentHash[2];
 
   InputFileEntry(FileEntryRef File) : File(File) {}
+
+  void trySetContentHash(Preprocessor &PP,
+                         std::optional<llvm::MemoryBufferRef> MemBuff) {
+    if (!PP.getHeaderSearchInfo()
```

Comment:
How about making the memory buffer a lambda so we can avoid re-opening the file if `ValidateASTInputFilesContent` is false?